src/controllers/updateTokenIcon.ts

```diff
@@ -0,0 +1,165 @@
+};
+
+const getVerifiedContract = async (
+  contractId
+) => {
+  const contract = await findVerifiedContract(
+    toChecksumAddress(contractId),
+  );
+  console.log(contract)
+  ensure(!!contract, 'Contract does not exist');
+  return contract;
+};
+
+const isValidSignature = (signedMessage, signature, address) => {
+  const publicKey = decodeAddress(address);
+  const hexPublicKey = u8aToHex(publicKey);
+
+  return signatureVerify(signedMessage, signature, hexPublicKey).isValid;
+};
+
+function generateSHA256Hash(inputString) {
+  const hash = crypto.createHash('sha256');
+  hash.update(inputString);
+  return hash.digest('hex');
+}
+
+const deleteNonceById=async(
+  id:string
+)=>{
+  try {
+    const nonceLookup = await uploadTokenIconRepository.findByPk(id);
+    if(nonceLookup){
+      await uploadTokenIconRepository.destroy({where:{address:id}})
+    } 
+  } catch (error) {
+  } 
+}
+
+export const generateNonce = async (
+  req: AppRequest<any>,
+  res: Response,
+) => {
+  const signerAddress = req.body['signerAddress'];
+  if(signerAddress.length === 0 || signerAddress === undefined) res.status(403).send("signerAddress not found");
+  
+  //generating nonce
+  const min = 0;
+  const max = 999;
+  const nonce = Math.floor(Math.random() * (max - min + 1)) + min;
+
+  // saving new nonce to db
+  try {
+    await uploadTokenIconRepository.upsert({
+      address:signerAddress,
+      nonce:nonce,
+    })
+  } catch (err: any) {
+    console.error(err);
+  }
+  res.status(200).send({
+    'nonce':nonce.toString(),
+  });
+}
+
+export const uploadTokenIcon = async (
+    req: AppRequest<any>,
+    res: Response,
+  ) => {
+    // extracting all the data from the req
+    const contractAddress = toChecksumAddress(req.body['contractAddress']);
+    const fileData = req.body['fileData'];
+    const uploadTimestamp = fileData['timestamp'];
+    const file = fileData['fileBase64'];
+    const signature = req.body['signature'];
+    const signerAddress = req.body['signerAddress'];
+    const fileHash = req.body['fileHash'];
+    const imageSizeInKB = Buffer.from(file, 'base64').length/ 1024;
+
+    // if image is more than 500KB return
+    if(imageSizeInKB>500){
+      res.status(403).send("image too big");
+    }
+
+    // obtaining file hash and comparing to check if it is same file
+    const calculatedFileHash = generateSHA256Hash(JSON.stringify(fileData));
+    if(fileHash !== calculatedFileHash){
+      res.status(403).send("different file sent");
+    }
+
+    // fetch the generated nonce
+    const uploadTokenIconResponse = await uploadTokenIconRepository.findByPk(signerAddress);
+    const nonce = uploadTokenIconResponse?.dataValues.nonce
+    await deleteNonceById(signerAddress);
+
+    // checking validity of signature
+    if(!isValidSignature([nonce?.toString()],signature,signerAddress)){
+      res.status(403).send('invalid signature')
+    }
+
+    // does contract have iconUri function
+    try {
```

Comment:
We still need to check if the addressSigner is the contract owner, should be something like:

  const abi = [
    "function owner() view returns (address)",
    "function iconUri() view returns (string)",
  ];
  const provider = getProvider()
  const contract = new Contract(contractAddress, abi, provider as any);

    try {
      const iconUri = await contract.iconUri()
      if(iconUri) res.status(403).send("icon already exists");
    } catch (error) {}

    try {
      const owner = await contract.owner()
      if(owner && owner !== ethers.constants.AddressZero && owner !== signerAddress) res.status(403).send("not contract owner");
    } catch (error) {}

Reply:
The try-catch block for owner call should not be nested in the try-catch block for iconUri call. If `contract.iconUri()` fails the error will be caught and `contract.owner()` will never be called.